File: leetcode/Unique_Binary_Search_Trees.java
public class Solution {
    int[] cache = new int[15];
    
    int solve(int n) {
        if (cache[n] != -1)
            return cache[n];
        cache[n] = 0;
        for (int l = 0; l <= n - 1; ++l)
            cache[n] += solve(l) * solve(n - 1 - l);
        return cache[n];
    }
    
    public int numTrees(int n) {
        for (int i = 0; i < 15; ++i)
            cache[i] = -1;
        cache[0] = cache[1] = 1;
        return solve(n);
    }
}

File: leetcode/Unique_Binary_Search_Trees.cpp
class Solution {
public:
    int numTrees(int n) {
        vector<int> cache;
        cache.push_back(1);  // n == 0
        cache.push_back(1);  // n == 1
        for (int i = 2; i <= n; ++i) {
            int sum = 0;
            for (int j = 0; j < i; ++j)
                sum += cache[j] * cache[i - 1 - j];
            cache.push_back(sum);
        }
        return cache[n];
    }
};
